magento/pwa-studio
# [PWA-2228] Shimmer loader for dynamic blocks

## Description

Add shimmer loader on dynamic blocks to avoid layout shift.
Shimmer loaders will only be rendered if minHeight is specified on the row containing a dynamic block
This gives flexibility to users to display a shimmer or not (let's say a dynamic block wouldn't render most of the time, we probably don't want to display a shimmer)

Following page showcase different combinations 
Use browser speed throttling if you can't see them.
https://jeremycharron-o4saqja-4txwfa4z2sor6.eu-4.magentosite.cloud/dynamic-block-shimmers

HTML code of CMS Page in PWA-2228 ticket

## Related Issue

Closes PWA-2228

## Acceptance

### Verification Stakeholders

Project maintainers

### Specification




## Verification Steps


#### Test scenario(s) for direct fix/feature

1. Add a row with a minHeight, then add a dynamic block inside. Verify that shimmer is rendered.
2. Add a row with a minHeight, then add 2 dynamic blocks inside. Verify that only one shimmer is displayed
3. Add a row with a minHeight, then add a dynamic block with conditions that cause the block not to render on front-end. Verify that shimmer still renders (this is expected), and white-space is removed when empty content is loaded.
4. Add a row, then add dynamic block inside. Verify that no shimmer is rendered (current Develop behavior)
5. Add a row with a minHeight, then add a column and add a dynamic block inside. Verify that shimmer is displayed

#### Test scenario(s) for any existing impacted features/areas

1. Verify that new and existing Jest tests pass
2. Verify that Cypress test passes

#### Test scenario(s) for any Magento Backend Supported Configurations

N/A

#### Is Browser/Device testing needed?

Yes, multiple browsers and devices.

#### Any ad-hoc/edge case scenarios that need to be considered?

Should be 100% compatible with recently added Personalized Sample data content (https://github.com/magento-commerce/venia-sample-data-modules-ee).

## Screenshots / Screen Captures (if appropriate)
Using page shared above, 

Develop without minHeight :
![cls-before-2](https://user-images.githubusercontent.com/25036022/154689875-f60f0abd-254f-462e-86a0-6a8944261708.png)

jcharron/PWA-2228 with minHeight :
![CLS](https://user-images.githubusercontent.com/25036022/154689934-d234404f-798e-4ed3-9702-d193d580a030.png)

Only the 4th test gets pushed, which is normal.
![cls-2](https://user-images.githubusercontent.com/25036022/154690056-8d21f2c2-b9a5-40fc-abba-e30b5fb8ec4c.png)

## Breaking Changes (if any)

No breaking change

## Checklist



-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.
